### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by adamdruppe**
As I go through building my ddoc killer, I'm noticing a lot of deficiencies in the doc sources, some big, some small.

std.algorithm.searching had a typo, for example, leading to a broken link. std.base64 had a LINK2 title with commas in it, leading to cut off text.

Several of them have LINK2 modules, which breaks anywhere except /phobos (even ddox has these broken links).

Many of the examples don't say what they are trying to do or don't show enough information to actually be useful.

Functions don't refer to each other when they obviously could (obvious to someone who knows Phobos already that is, the new reader would have no idea). In general, browsing Phobos docs, it is easy to lose the forest for the trees. There's few examples of how to integrate modules.

There's an anti-pattern of two macros always appearing together, like `$(M1 $(M2 ...))`. That's silly, if M1 should _always_ be applied to M2, just change M2's definition.

A great many FAQs in the support arena are not answered in the docs. I am adding a new section called "Diagnostics" with sample compiler error messages and showing how to fix them.

New users go to std.path to see what is in the current path, but that is actually in std.file. Let's have them link to each other.

Some functions create their own `$(RED Important:) ...` constructs. I propose changing them to be a well-defined semantic macro like `$(PITFALL)`.

In a few places, I also add a blank line. This should be irrelevant to phobos, but it gives my doc generator a natural place to insert a table of contents after a brief introduction. The first line in ddoc is supposed to be a summary. Some functions lack this. Some write it with the assumption that it will be read immediately preceding the rest of the comment, making them read nonsensically if extracted into said table of contents. I reword a few of these.

I rearranged the `toStringz` overloads to put the one with the simpler signature first. Then a user browsing down will see that simple one as the entry point and the other one as an alternative. A small thing that has a bigger source diff than it really is, but makes things a bit more readable for the first-time browser.

I have a lot more in store. I'm willing to compromise a little to maintain compatibility with ddoc while still getting content enhancements merged, though my main goal here is to make better docs.

### CODE REVIEW

File: std/algorithm/searching.d [hunk 1]
```diff
@@ -2598,7 +2609,7 @@ if (isForwardRange!R1 && isForwardRange!R2)
     }
 }
 
-///
+/// This example shows the difference in return values of $(LREF findSkip), $(LREF findSplitBefore), and $(LREF findSplitAfter).
```
Comment (JackStouffer) [inline std/algorithm/searching.d]: line length

File: std/base64.d [hunk 2]
```diff
@@ -76,7 +75,32 @@ unittest
...
- * See $(D $(LREF Base64Impl)) for a description of available methods.
+ * See $(LREF Base64Impl) for a description of available methods.
+ *
+ * Diagnostics:
+ *
+ * If you try to encode a string, or some other non-$(D ubyte[]) data, you are
+ * likely to see a mess of an error message from the compiler that starts with
+ * a line like this:
+ *
+ * `a.d(6): Error: template std.base64.Base64Impl!('+', '/', '=').encode cannot deduce function from argument types !()(string), candidates are:`
```
Comment (JackStouffer) [inline std/base64.d]: ditto

File: std/conv.d [hunk 3]
```diff
@@ -1,7 +1,7 @@
 // Written in the D programming language.
 
 /**
-A one-stop shop for converting values from one type to another.
+A one-stop shop for converting values from one type to another. Typical usage: `$(LREF to)!SomeType(some_variable);`
```
Comment (JackStouffer) [inline std/conv.d]: ditto

File: std/string.d [hunk 4]
```diff
@@ -73,6 +73,7 @@ $(TR $(TDNW Miscellaneous)
     )
 )))
 
+
```
Comment (JackStouffer) [inline std/string.d]: unnecessary line space

File: std/string.d [hunk 5]
```diff
@@ -244,12 +250,32 @@ inout(char)[] fromStringz(inout(char)* cString) @nogc @system pure nothrow {
...
 
-    $(RED Important Note:) When passing a $(D char*) to a C function, and the C
+    $(PITFALL When passing a $(D char*) to a C function, and the C
     function keeps it around for any reason, make sure that you keep a
     reference to it in your D code. Otherwise, it may become invalid during a
     garbage collection cycle and cause a nasty bug when the C code tries to use
-    it.
+    it.)
   +/
+immutable(char)* toStringz(in string s) @trusted pure nothrow
```
Comment (JackStouffer) [inline std/string.d]: Why was this moved?

Comment (adamdruppe) [inline std/string.d]: I explained this in my PR description.

"I rearranged the  toStringz  overloads to put the one with the simpler signature first. Then a user browsing down will see that simple one as the entry point and the other one as an alternative. A small thing that has a bigger source diff than it really is, but makes things a bit more readable for the first-time browser."

File: std/string.d (hunk 4 above)
Comment (adamdruppe) [inline std/string.d]: On Thu, Dec 31, 2015 at 05:26:23PM -0800, Jack Stouffer wrote:
> unnecessary line space

I explained this in the opening comment too.

"In a few places, I also add a blank line. This should be irrelevant to phobos, but it gives my doc generator a natural place to insert a table of contents after a brief introduction. The first line in ddoc is supposed to be a summary. Some functions lack this. Some write it with the assumption that it will be read immediately preceding the rest of the comment, making them read nonsensically if extracted into said table of contents. I reword a few of these"

**Comment by CyberShadow** [comment]
Please try to use more descriptive commit messages; none of the ones you wrote are very informative.

**Comment by CyberShadow** [comment]
> There's an anti-pattern of two macros always appearing together, like $(M1 $(M2 ...)). That's silly, if M1 should always be applied to M2, just change M2's definition.

I assume you mean `LNAME` and `D`? Not if the two are semantically distinct things - such as linking to things and marking what is D code. But even then, changing M2's definition is wrong on many levels, you should define a new marso instead.
 
Is `LNAME`/`LNAME2` really used everywhere exclusively together with `D`?

**Comment by adamdruppe** [comment]
On Thu, Dec 31, 2015 at 07:23:45PM -0800, Vladimir Panteleev wrote:
> I assume you mean `LNAME` and `D`?

No, I don't even know what LNAME is. It doesn't show up in a grep either.

Oh there it is, in the druntime changelog...

But the one I had in mind was: $(D $(LREF appender))

If an LREF should always be rendered in code font, just make it
output a code font. And it shouldn't be always used, why is it
used here?

**Comment by CyberShadow** [comment]
Yep, sorry, LREF not LNAME.

LREF is short for "local reference"... nothing about its name implies that its contents will be in code font. Do I really need to explain why changing the meaning of macros is bad?

**Comment by adamdruppe** [comment]
On Thu, Dec 31, 2015 at 07:44:07PM -0800, Vladimir Panteleev wrote:
> LREF is short for "local reference"... nothing about its name implies that its contents will be in code font.

Then why is it EVER in code font? On one of these pages, every
instance of it was wrapped in $(D). Why?

> Do I really need to explain why changing the meaning of macros is bad?

This isn't the meaning, it seems to just be about a styling issue,
choosing a font.

**Comment by CyberShadow** [comment]
> Then why is it EVER in code font? On one of these pages, every instance of it was wrapped in $(D). Why?

I don't understand what you mean - clearly the answer is because every instance of it referred to a code identifier.

Anyway, I just looked at it (sorry I'm not home right now so I'm limited in doing such research), and didn't realize that LREF was never used on 'dlang.org`, only Phobos and always referring to symbols.

However, there are cases like: `$(D GC.BlkAttr.$(LREF NO_MOVE))`

To avoid regressions in HTML output, we would need to merge PRs introducing the LREF change simultaneously with all PRs in Phobos/Druntime that fix LREF references. It might be easier to do this by introducing a new macro first, then changing content to use that macro, then maybe removing the old macro. This also brings no confusion of what `LREF` means at any point in time.

**Comment by adamdruppe** [comment]
On Thu, Dec 31, 2015 at 10:36:43PM -0800, Vladimir Panteleev wrote:
> However, there are cases like: `$(D GC.BlkAttr.$(LREF NO_MOVE))`

That should just be $(LREF GC.BlkAttr.NO_MOVE) because the *whole* symbol is a local reference. Since DDOC_ANCHOR is defined to put a leading dot in front, LREF should do the same thing, so it continues to work on everything.

All those instances should be replaced. It was never actually correct to link to just one part of the name. If a global symbol was ever introduced with the same name, it would break that link!

> To avoid regressions in HTML output, we would need to merge PRs introducing the LREF change simultaneously with all PRs in Phobos/Druntime that fix LREF references.

If it is a regression to add a html class to a link, this process is more broken than I realized.

> This also brings no confusion of what `LREF` means at any point in time.

LREF means "local reference" and since you're documenting code, a local reference also happens to be a code identifier, so the css may style it as one.

**Comment by CyberShadow** [comment]
> That should just be $(LREF GC.BlkAttr.NO_MOVE) because the *whole* symbol is a local reference.

OK, just saying that such instances will need to be updated.

> If it is a regression to add a html class to a link, this process is more broken than I realized.

Not sure what process you refer to, but I thought there's more to it than a CSS class.

Anyway, I just looked at LREF again and realized that it's already defined as `<a href="#$1">$(D $1)</a>`, so the `$(D ...)` is indeed redundant. Sorry for the noise.

**Comment by adamdruppe** [comment]
On Fri, Jan 01, 2016 at 11:59:09AM -0800, Vladimir Panteleev wrote:
> OK, just saying that such instances will need to be updated.

Right, I did a grep in Phobos and pushed the commit here too
(along with other things, I'm just editing docs as I see them
in the one branch)

**Comment by adamdruppe** [comment]
So I'm continuing to push almost random changes here mostly just so you can see what I've been doing with my fork.

Between the pretty blue links in the beautifully formatted navigation bar telling me what I haven't looked at yet and the ease of just editing a file here without spending a lot of brain power on ddoc strangeness (most of this remains compatible but some of it isn't) and just not caring about the auto tester's irrelevant nitpicking, I'm able to make quite a few nice changes.

**Comment by greenify** [comment]
(ping) the longer you wait, the more conflicts you might get ;-)

**Comment by adamdruppe** [comment]
I'm periodically pulling and rebasing to keep my changes up to date. However, there's mixes of ddoc compatible and non-ddoc changes in here now too.

**Comment by greenify** [comment]
Ah I thought you wanted to merge those typos and better descriptions. My fault :/

Is there an easy way to separate the compatible changes?

**Comment by adamdruppe** [comment]
On Fri, Mar 04, 2016 at 12:12:34PM -0800, Seb wrote:
> Is there an easy way to separate the compatible changes? 

I think going through the diff line by line will do it, but I still
have a few other things to do.

**Comment by greenify** [comment]
> Several of them have LINK2 modules, which breaks anywhere except /phobos (even ddox has these broken links).

This is also addressed in #3926.

**Comment by adamdruppe** [comment]
On Wed, Mar 30, 2016 at 01:19:04PM -0700, Seb wrote:
> > Several of them have LINK2 modules, which breaks anywhere except /phobos (even ddox has these broken links).
> 
> This is also addressed in #3926.

That's why I wrote the MREF macro!

**Comment by wilzbach** [comment]
So more than a year went by & apparently no one wanted to merge this big PR. Moreover, thanks to a couple of deprecations (e.g. removal of `XREF`) there are quite a bunch of conflicts now.

@adamdruppe I am now closing this with regrets due to inactivity, but I highly encourage you to resubmit this work - maybe split into a PR per module?

**Comment by adamdruppe** [comment]
On Sun, Dec 25, 2016 at 10:02:48PM -0800, Sebastian Wilzbach wrote:
> @adamdruppe I am now closing this with regrets due to inactivity, but I highly encourage you to resubmit this work - maybe split into a PR per module?

I'm doing it all on a private fork now anyway, I decided to just drop compatibility with vanilla ddoc and am no longer attempting to work
with it.
